File: Producto/frontend/src/lib/matchEngine.ts
// matchEngine.ts

export interface MatchScore {
  total: number;
  breakdown: {
    habilidades: number; // Max 40
    experiencia: number; // Max 20
    distancia: number;   // Max 20
    disponibilidad: number; // Max 20
    calificacion: number; // Bonus Max 10
  };
}

// Helpers
function calculateDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
  if (!lat1 || !lon1 || !lat2 || !lon2) return -1;
  const R = 6371; // km
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a = 
    Math.sin(dLat/2) * Math.sin(dLat/2) +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * 
    Math.sin(dLon/2) * Math.sin(dLon/2); 
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a)); 
  return R * c;
}

/**
 * Calcula el puntaje de compatibilidad entre un trabajador y una oferta/empresa.
 * @param workerProfile  Perfil del trabajador
 * @param companyProfile Perfil de la empresa
 * @param jobOffer       Oferta de trabajo
 * @param callerPlan     Plan del usuario que realiza la búsqueda ('FREE' | 'PREMIUM' | 'ENTERPRISE')
 */
export function calculateMatchScore(workerProfile: any, companyProfile: any, jobOffer: any, callerPlan: string = 'FREE'): MatchScore {
  const isPremium = callerPlan === 'PREMIUM' || callerPlan === 'ENTERPRISE';

  let score: MatchScore = {
    total: 0,
    breakdown: { habilidades: 0, experiencia: 0, distancia: 0, disponibilidad: 0, calificacion: 0 }
  };

  // Safe parsing of skills JSON
  const wSkills = workerProfile?.skills ? (typeof workerProfile.skills === 'string' ? JSON.parse(workerProfile.skills) : workerProfile.skills) : {};
  const cSkills = companyProfile?.skills ? (typeof companyProfile.skills === 'string' ? JSON.parse(companyProfile.skills) : companyProfile.skills) : {};

  // 1. HABILIDADES (40%)
  let habScore = 0;
  
  // A. Característica (Trabajador) vs Habilidad Valorada (Empresa) [Max 20%]
  const wCaract = wSkills.caracteristica || '';
  const cHabil = cSkills.habilidadValorada || '';
  
  const charactMap: Record<string, string> = {
    'Responsable': 'Responsabilidad',
    'Rápido': 'Rapidez',
    'Líder': 'Liderazgo',
    'Comunicativo': 'Comunicación',
    'Analítico': 'Resolución de problemas' // Proxy
  };

  if (wCaract && cHabil) {
    if (charactMap[wCaract] === cHabil || wCaract === cHabil) {
      habScore += 20;
    } else {
      habScore += 10; // Partial match just for having something
    }
  }

  // B. Ambiente (Trabajador) vs Ritmo (Empresa) [Max 20%]
  const wAmbiente = wSkills.ambiente || '';
  const cRitmo = cSkills.ritmo || '';
  
  if (wAmbiente && cRitmo) {
    if (wAmbiente === cRitmo) {
      habScore += 20;
    } else if (
      (wAmbiente === 'Alta presión' && cRitmo === 'Muy dinámico') ||
      (wAmbiente === 'Trabajo en equipo' && cRitmo === 'Estructurado')
    ) {
      habScore += 15;
    } else {
      habScore += 5;
    }
  } else if (wSkills.habilidades && wSkills.habilidades.length > 0) {
     // Fallback: If they selected skills, give them some points
     habScore += 10;
  }
  
  score.breakdown.habilidades = Math.min(40, habScore);

  // 2. EXPERIENCIA (20%)
  // Since we don't have an explicit experience field, we check description length/keywords
  let expScore = 0;
  const desc = (workerProfile?.description || '').toLowerCase();
  if (desc.length > 10) expScore += 10;
  if (desc.includes('experiencia') || desc.includes('años') || desc.includes('trabajé')) {
    expScore += 10;
  }
  score.breakdown.experiencia = expScore;

  // 3. DISTANCIA (20%) — Diferenciada por plan
  // FREE:    ≤5 km = 20 pts | >5 km = 0 pts (fuera de rango gratuito)
  // PREMIUM: ≤5 km = 20 pts | ≤10 km = 15 pts | ≤15 km = 10 pts | >15 km = 5 pts
  let distScore = 0;
  const targetLat = jobOffer?.latitude || companyProfile?.latitude;
  const targetLng = jobOffer?.longitude || companyProfile?.longitude;
  
  const dist = calculateDistance(workerProfile?.latitude, workerProfile?.longitude, targetLat, targetLng);
  
  if (dist === -1) {
    distScore = 10; // Default if no coords available
  } else if (isPremium) {
    // PREMIUM: escalonado completo hasta 15 km
    if (dist <= 5) {
      distScore = 20;
    } else if (dist <= 10) {
      distScore = 15;
    } else if (dist <= 15) {
      distScore = 10;
    } else {
      distScore = 5;
    }
  } else {
    // FREE: solo puntaje dentro de 5 km, fuera = 0
    if (dist <= 5) {
      distScore = 20;
    } else {
      distScore = 0;
    }
  }
  score.breakdown.distancia = distScore;

  // 4. DISPONIBILIDAD (20%)
  let dispScore = 0;
  const wPref = wSkills.preferencia || '';
  // If the job description or title implies part time / full time, we can match
  const jobTitleDesc = ((jobOffer?.titulo || '') + ' ' + (jobOffer?.descripcion || '')).toLowerCase();
  
  if (wPref) {
    if (jobTitleDesc.includes(wPref.toLowerCase())) {
      dispScore = 20;
    } else {
      dispScore = 15; // They provided availability, assume partial match
    }
  } else {
    dispScore = 10; // Default
  }
  score.breakdown.disponibilidad = dispScore;

  // 5. CALIFICACIÓN (BONUS +10)
  // Utilizamos calificación real del perfil
  const workerRating = workerProfile?.rating || 0;
  let califScore = 0;
  if (workerRating >= 4.8) {
    califScore = 10;
  } else if (workerRating >= 4.5) {
    califScore = 7;
  } else if (workerRating >= 4.0) {
    califScore = 4;
  }
  score.breakdown.calificacion = califScore;

  // Total base
  score.total = score.breakdown.habilidades + score.breakdown.experiencia + score.breakdown.distancia + score.breakdown.disponibilidad + score.breakdown.calificacion;
  
  // PREMIUM BONUS: +20% al puntaje total
  if (isPremium) {
    score.total = Math.round(score.total * 1.20);
  }

  // Ensure it doesn't exceed 100
  score.total = Math.min(100, Math.round(score.total));

  return score;
}

File: Producto/android-app/app/src/main/java/cl/catchgo/app/core/util/MatchEngine.kt
package cl.catchgo.app.core.util

import cl.catchgo.app.data.remote.dto.JobOfferDto
import cl.catchgo.app.data.remote.dto.ProfileRemoteDto
import org.json.JSONObject
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

object MatchEngine {

    data class MatchScore(
        val total: Int,
        val habilidades: Int,
        val experiencia: Int,
        val distancia: Int,
        val disponibilidad: Int,
        val calificacion: Int
    )

    private val charactMap = mapOf(
        "Responsable" to "Responsabilidad",
        "Rápido" to "Rapidez",
        "Líder" to "Liderazgo",
        "Comunicativo" to "Comunicación",
        "Analítico" to "Resolución de problemas"
    )

    fun calculate(
        workerProfile: ProfileRemoteDto,
        companyProfile: ProfileRemoteDto,
        jobOffer: JobOfferDto?
    ): MatchScore {
        val wSkills = parseSkills(workerProfile.skills)
        val cSkills = parseSkills(companyProfile.skills)

        val habilidades = calcHabilidades(wSkills, cSkills)
        val experiencia = calcExperiencia(workerProfile.description)
        val distancia = calcDistancia(
            workerProfile.latitude, workerProfile.longitude,
            jobOffer?.latitude ?: companyProfile.latitude,
            jobOffer?.longitude ?: companyProfile.longitude
        )
        val disponibilidad = calcDisponibilidad(wSkills, jobOffer)
        val calificacion = calcCalificacion(workerProfile.rating)

        val total = min(100, habilidades + experiencia + distancia + disponibilidad + calificacion)
        return MatchScore(total, habilidades, experiencia, distancia, disponibilidad, calificacion)
    }

    private fun parseSkills(json: String?): JSONObject {
        if (json.isNullOrBlank()) return JSONObject()
        return runCatching { JSONObject(json) }.getOrElse { JSONObject() }
    }

    private fun calcHabilidades(wSkills: JSONObject, cSkills: JSONObject): Int {
        var score = 0
        val wCaract = wSkills.optString("caracteristica", "")
        val cHabil = cSkills.optString("habilidadValorada", "")
        if (wCaract.isNotEmpty() && cHabil.isNotEmpty()) {
            score += if (charactMap[wCaract] == cHabil || wCaract == cHabil) 20 else 10
        }
        val wAmbiente = wSkills.optString("ambiente", "")
        val cRitmo = cSkills.optString("ritmo", "")
        if (wAmbiente.isNotEmpty() && cRitmo.isNotEmpty()) {
            score += when {
                wAmbiente == cRitmo -> 20
                (wAmbiente == "Alta presión" && cRitmo == "Muy dinámico") ||
                        (wAmbiente == "Trabajo en equipo" && cRitmo == "Estructurado") -> 15
                else -> 5
            }
        } else {
            val habilidades = wSkills.optJSONArray("habilidades")
            if (habilidades != null && habilidades.length() > 0) score += 10
        }
        return min(40, score)
    }

    private fun calcExperiencia(description: String?): Int {
        val desc = description?.lowercase() ?: return 0
        var score = 0
        if (desc.length > 10) score += 10
        if (desc.contains("experiencia") || desc.contains("años") || desc.contains("trabajé")) score += 10
        return score
    }

    private fun calcDistancia(
        lat1: Double?, lon1: Double?,
        lat2: Double?, lon2: Double?
    ): Int {
        if (lat1 == null || lon1 == null || lat2 == null || lon2 == null) return 10
        val dist = haversineKm(lat1, lon1, lat2, lon2)
        return when {
            dist <= 10.0 -> 20
            dist <= 25.0 -> 10
            else -> 5
        }
    }

    private fun calcDisponibilidad(wSkills: JSONObject, jobOffer: JobOfferDto?): Int {
        val pref = wSkills.optString("preferencia", "")
        if (pref.isEmpty()) return 10
        val jobText = ((jobOffer?.titulo ?: "") + " " + (jobOffer?.descripcion ?: "")).lowercase()
        return if (jobText.contains(pref.lowercase())) 20 else 15
    }

    private fun calcCalificacion(rating: Double?): Int {
        return when {
            rating == null || rating == 0.0 -> 0
            rating >= 4.8 -> 10
            rating >= 4.5 -> 7
            rating >= 4.0 -> 4
            else -> 0
        }
    }

    private fun haversineKm(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val r = 6371.0
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = sin(dLat / 2) * sin(dLat / 2) +
                cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) *
                sin(dLon / 2) * sin(dLon / 2)
        return r * 2 * atan2(sqrt(a), sqrt(1 - a))
    }
}
